### PR TITLE
EIP-1559 tx pool support

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This pull request implements EIP-1559 compatible transaction pool with dual heap eviction ordering.
It is based on https://github.com/ethereum/go-ethereum/pull/22791
The eviction ordering scheme and the reasoning behind it is described here: https://gist.github.com/zsfelfoldi/9607ad248707a925b701f49787904fd6

### CODE REVIEW

File: core/tx_list.go [hunk 1]
```diff
@@ -279,15 +279,21 @@ func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Tran
...
-		// Have to ensure that the new gas price is higher than the old gas
-		// price as well as checking the percentage threshold to ensure that
+		thresholdFeeCap := aFeeCap.Div(aFeeCap, b)
+		thresholdTip := aTip.Div(aTip, b)
+
+		// Have to ensure that either the new fee cap or tip is higher than the
+		// old ones as well as checking the percentage threshold to ensure that
 		// this is accurate for low (Wei-level) gas price replacements
-		if old.GasPriceCmp(tx) >= 0 || tx.GasPriceIntCmp(threshold) < 0 {
+		if (old.FeeCapCmp(tx) >= 0 || tx.FeeCapIntCmp(thresholdFeeCap) < 0) ||
```
Comment (rjl493456442) [inline core/tx_list.go:295]: It can be problematic if both `FeeCap` and `Tip` are less than the old one. The check is skipped.

Comment (holiman) [reply core/tx_list.go:295]: IMO we could break out the 'trivial' checks to happen first, making this if statement smaller and avoiding arithmetics. 

```golang
	if old != nil {
		if old.FeeCapCmp(tx) >= 0 || old.TipCmp(tx) >= 0 { 
			return false, nil
		} 
		// thresholdFeeCap = oldFC * (100 + priceBump) / 100
		...
		if tx.FeeCapIntCmp(thresholdFeeCap) < 0) || tx.TipIntCmp(thresholdTip) < 0){
```

File: core/tx_pool.go [hunk 2]
```diff
@@ -540,13 +553,17 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	if pool.currentMaxGas < tx.Gas() {
 		return ErrGasLimit
 	}
+	// Ensure feeCap is less than or equal to tip.
+	if pool.eip1559 && tx.FeeCapIntCmp(tx.Tip()) < 0 {
+		return ErrTipAboveFeeCap
+	}
```
Comment (holiman) [inline core/tx_pool.go:559]: Is `pool.eip1559` needed here?

File: core/tx_pool_test.go [hunk 3]
```diff
@@ -87,6 +97,22 @@ func pricedDataTransaction(nonce uint64, gaslimit uint64, gasprice *big.Int, key
 	return tx
 }
 
+func dynamicFeeTx(nonce uint64, gaslimit uint64, gasFee *big.Int, tip *big.Int, key *ecdsa.PrivateKey) *types.Transaction {
+	tx := types.NewTx(&types.DynamicFeeTx{
```
Comment (holiman) [inline core/tx_pool_test.go:101]: You can use `types.SignNewTx(..)` instead

File: core/tx_pool.go (hunk 2 above)
Comment (zsfelfoldi) [reply core/tx_pool.go]: Not really, removed it

File: core/tx_pool_test.go (hunk 3 above)
Comment (zsfelfoldi) [reply core/tx_pool_test.go]: right, done

File: core/tx_list.go [hunk 4]
```diff
@@ -279,15 +279,23 @@ func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Tran
 	// If there's an older better transaction, abort
 	old := l.txs.Get(tx.Nonce())
 	if old != nil {
-		// threshold = oldGP * (100 + priceBump) / 100
+		if old.FeeCapCmp(tx) >= 0 || old.TipCmp(tx) >= 0 {
```
Comment (rjl493456442) [inline core/tx_list.go:282]: In this strategy, we can only replace the transaction with 10% feeCap bump and 10% tip bump. It might hurt the UX to some extend.

For example, if the user already specify a high enough feeCap but a barely enough tip. And he wants to speed up the inclusion, he has to bump the feeCap as well. If the basefee raises up, then the maximum possible payment is increased. 

Of course, bumping both factors is easiest. Not sure if it's worthwhile to introduce addition complexity

File: core/tx_pool.go [hunk 5]
```diff
@@ -595,10 +612,10 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (replaced bool, err e
 		return false, err
 	}
 	// If the transaction pool is full, discard underpriced transactions
-	if uint64(pool.all.Count()+numSlots(tx)) > pool.config.GlobalSlots+pool.config.GlobalQueue {
+	if uint64(pool.all.Slots()+numSlots(tx)) > pool.config.GlobalSlots+pool.config.GlobalQueue {
 		// If the new transaction is underpriced, don't accept it
 		if !isLocal && pool.priced.Underpriced(tx) {
-			log.Trace("Discarding underpriced transaction", "hash", hash, "price", tx.GasPrice())
+			log.Trace("Discarding underpriced transaction", "hash", hash, "price", tx.FeeCap())
```
Comment (rjl493456442) [inline core/tx_pool.go:618]: Please also add the `tx.Tip()` information into the log

File: core/tx_pool.go [hunk 6]
```diff
@@ -615,7 +632,7 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (replaced bool, err e
 		}
 		// Kick out the underpriced remote transactions.
 		for _, tx := range drop {
-			log.Trace("Discarding freshly underpriced transaction", "hash", tx.Hash(), "price", tx.GasPrice())
+			log.Trace("Discarding freshly underpriced transaction", "hash", tx.Hash(), "price", tx.FeeCap())
```
Comment (rjl493456442) [inline core/tx_pool.go:635]: Please also add the tx.Tip() information into the log

**Comment by holiman** [comment]
Rebased + fixed rebase problems

**Comment by holiman** [comment]
`GasPriceCmp` is now unused, and should be removed. `GasPriceIntCmp` is used by the GPO, but I guess the GPO should use something else, and we should remove that method too.

File: core/tx_pool.go (hunk 5 above)
Comment (zsfelfoldi) [reply core/tx_pool.go]: done

File: core/tx_pool.go (hunk 6 above)
Comment (zsfelfoldi) [reply core/tx_pool.go]: done

File: core/tx_list.go (hunk 4 above)
Comment (zsfelfoldi) [reply core/tx_list.go:282]: This has been discussed at the 1559 dev channel, right now this replacement condition seems to be the best choice.

File: core/tx_pool.go [hunk 7]
```diff
@@ -540,13 +555,17 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	if pool.currentMaxGas < tx.Gas() {
 		return ErrGasLimit
 	}
+	// Ensure feeCap is less than or equal to tip.
```
Comment (rjl493456442) [inline core/tx_pool.go:558]: At L540, Please also reject the 1559 txs if the fork is not happened yet.
Although during the transaction processing, the 1559 transactions will be filtered out, but would be nice to reject it in the first place.

**Comment by holiman** [comment]
This is now running on `bootnode-aws-eu-central-1-001`

File: core/tx_pool_test.go [hunk 8]
```diff
@@ -97,6 +122,16 @@ func setupTxPool() (*TxPool, *ecdsa.PrivateKey) {
 	return pool, key
```
Comment (MariusVanDerWijden) [inline core/tx_pool_test.go:122]:
```suggestion
	return setupTxPoolWithConfig(params.TestChainConfig)
```

File: core/tx_pool.go (hunk 7 above)
Comment (zsfelfoldi) [reply core/tx_pool.go:562]: done

File: core/tx_pool_test.go (hunk 8 above)
Comment (zsfelfoldi) [reply core/tx_pool_test.go]: done

**Comment by holiman** [comment]
Baikal sealer now running `instance=Geth/v1.10.4-unstable-dd5f811b-20210527/linux-amd64/go1.16.4`

File: core/tx_list.go [hunk 9]
```diff
@@ -552,12 +593,32 @@ func (l *txPricedList) Discard(slots int, force bool) (types.Transactions, bool)
...
 	}, false, true) // Only iterate remotes
-	heap.Init(l.remotes)
+	heap.Init(&l.urgent)
+
+	// balance out the two heaps by moving the worse half of transactions into the
+	// floating heap
+	// Note: Discard would also do this before the first eviction but Reheap can do
+	// is more efficiently. Also, Underpriced would work suboptimally the first time
+	// if the floating queue was empty.
+	floatingCount := len(l.urgent.list) * floatingRatio / (urgentRatio + floatingRatio)
```
Comment (karalabe) [inline core/tx_list.go:610]: Shouldn't floatingCount be only the excess above the urgent cap? E.g. if we have 4+1K txs configured, we should only overflow if above 4K. Currently if I have 10txs, I'll overflow 2, which imho doesn't make sense.

File: core/tx_list.go [hunk 10]
```diff
@@ -529,21 +555,36 @@ func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
...
 func (l *txPricedList) Discard(slots int, force bool) (types.Transactions, bool) {
 	drop := make(types.Transactions, 0, slots) // Remote underpriced transactions to drop
-	for len(*l.remotes) > 0 && slots > 0 {
-		// Discard stale transactions if found during cleanup
-		tx := heap.Pop(l.remotes).(*types.Transaction)
-		if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
-			l.stales--
-			continue
+	for slots > 0 {
+		if len(l.urgent.list)*floatingRatio > len(l.floating.list)*urgentRatio || floatingRatio == 0 {
```
Comment (karalabe) [inline core/tx_list.go:559]: I don't think it's a good idea to rebalance things if they are not full. We should keep shoving txs into the urgent queue until it fills up and only then overflow into the floating one. Am I missing something perhaps? I don't see any upside, only downsides to moving txs into the floating queue while there's room in the urgent queue.

File: core/tx_pool.go [hunk 11]
```diff
@@ -83,6 +83,10 @@ var (
 	// than some meaningful limit a user might use. This is not a consensus error
 	// making the transaction invalid, rather a DOS protection.
 	ErrOversizedData = errors.New("oversized data")
+
+	// ErrTipAboveFeeCap is a sanity error to ensure no one is able to specify a
+	// transaction with a tip higher than the total fee cap.
+	ErrTipAboveFeeCap = errors.New("tip higher than fee cap")
```
Comment (karalabe) [inline core/tx_pool.go:89]: Please use `tip cap`, not `tip`. There's no notion of user configurable tip.

File: core/tx_list.go [hunk 12]
```diff
@@ -521,7 +521,11 @@ func (l *txPricedList) Removed(count int) {
 // Underpriced checks whether a transaction is cheaper than (or as cheap as) the
 // lowest priced (remote) transaction currently being tracked.
 func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
-	return l.underpricedFor(&l.urgent, tx) && l.underpricedFor(&l.floating, tx)
+	// Note: with two queues, being underpriced is defined as being worse than the worst item
+	// in all non-empty queues if there is any. If both queues are empty then nothing is underpriced.
+	return (l.underpricedFor(&l.urgent, tx) || len(l.urgent.list) == 0) &&
```
Comment (karalabe) [inline core/tx_list.go:526]: I don't get these special cases. Why return "underpriced" if a queue is empty?

File: core/tx_list.go (hunk 10 above)
Comment (zsfelfoldi) [reply core/tx_list.go:560]: In `txPricedList` the definition of full is that items are being discarded. The `urgentRatio : floatingRatio` thing is internal. If I wanted to enforce absolute limits here then I would need to deal with slot counting because that's how the global limit is calculated in the tx pool. I don't think that complexity is necessary. I think this is a clearly defined behavior: if we discard for any reason at any point, we discard from the two queues alternating in a 4:1 ratio (before which a rebalancing is automatically performed so that everyone is where they should be). We don't assume anything very specific about the caller but this is a safe and rational strategy for any meaningful use of `txPricedList`.

File: core/tx_list.go (hunk 9 above)
Comment (zsfelfoldi) [reply core/tx_list.go:611]: See below

File: core/tx_list.go (hunk 12 above)
Comment (zsfelfoldi) [reply core/tx_list.go:526]: See the comment above. We don't report underpriced if a queue is empty, we just don't automatically consider it non-underpriced. The last line fixes the condition that I initially broke, so if there are no non-empty queues then obviously nothing is underpriced.

File: core/tx_pool.go (hunk 11 above)
Comment (zsfelfoldi) [reply core/tx_pool.go:89]: Is this now the "official term"? Or should it be? It was called "tip" originally, even though it's a max (at least by some people). If we change it then maybe we should converge to some canonized terms.

File: core/tx_list.go [hunk 13]
```diff
@@ -529,21 +556,36 @@ func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
...
 		}
-		// Non stale transaction found, discard it
-		drop = append(drop, tx)
-		slots -= numSlots(tx)
 	}
 	// If we still can't make enough room for the new transaction
 	if slots > 0 && !force {
 		for _, tx := range drop {
-			heap.Push(l.remotes, tx)
+			heap.Push(&l.urgent, tx)
```
Comment (karalabe) [inline core/tx_list.go:588]: This is incorrect. The transaction most probably originated from the floating heap, not the urgent heap. Furthermore, we have done certain urgent -> floating heap movements, which should also be reverted in case we can't discard the required number of slots.

File: core/tx_list.go (hunk 12 above)
Comment (karalabe) [reply core/tx_list.go:526]: I still don't get it why we need to care if a queue is empty. If it's empty, there's room to put the tx in there, so it's not underpriced. What am I missing?

File: core/tx_list.go (hunk 13 above)
Comment (zsfelfoldi) [reply core/tx_list.go:588]: With reverting urgent->floating movements you are theoretically right, I'll think about a clean and simple way to handle this.Putting everything in urgent is good though, we can always add everything to urgent first and balance the queues out before discarding anything.

File: core/tx_list.go (hunk 12 above)
Comment (zsfelfoldi) [reply core/tx_list.go:526]: It's just a corner case that's not going to happen normally but for example a test fails if I don't have this check because it uses a low limit and nothing fits in the floating queue which would make nothing underpriced, ever. I think this condition is good enough for those corner cases while in normal case where 1/4 of the total remotes count is always >= 1 it doesn't matter.
